File: java/src/main/java/com/ollang/sdk/resources/CustomInstructions.java
package com.ollang.sdk.resources;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.ollang.sdk.OllangClient;

/** Manage custom translation instructions. */
public class CustomInstructions {

  private final OllangClient client;

  public CustomInstructions(OllangClient client) {
    this.client = client;
  }

  public JsonElement list() {
    return client.get("/integration/custom-instructions");
  }

  /** Creates a custom instruction. {@code description} may be {@code null}. */
  public JsonElement create(String key, String value, String description) {
    JsonObject body = new JsonObject();
    body.addProperty("key", key);
    body.addProperty("value", value);
    if (description != null) {
      body.addProperty("description", description);
    }
    return client.post("/integration/custom-instructions", body);
  }

  /** Updates a custom instruction. Any of {@code key}, {@code value}, {@code description} may be {@code null} to leave it unchanged. */
  public JsonElement update(String instructionId, String key, String value, String description) {
    JsonObject body = new JsonObject();
    if (key != null) {
      body.addProperty("key", key);
    }
    if (value != null) {
      body.addProperty("value", value);
    }
    if (description != null) {
      body.addProperty("description", description);
    }
    return client.patch("/integration/custom-instructions/" + instructionId, body);
  }

  public JsonElement delete(String instructionId) {
    return client.delete("/integration/custom-instructions/" + instructionId);
  }

  public JsonElement suggestions() {
    return client.get("/integration/custom-instructions/suggestions");
  }
}

File: python/src/ollang/resources/custom_instructions.py
from typing import Any, Dict, List, Optional

from .._client import OllangClient


class CustomInstructions:
    """Manage custom translation instructions."""

    def __init__(self, client: OllangClient):
        self._client = client

    def list(self) -> List[Dict[str, Any]]:
        return self._client.get("/integration/custom-instructions")

    def create(self, key: str, value: str, description: Optional[str] = None) -> Dict[str, Any]:
        body: Dict[str, Any] = {"key": key, "value": value}
        if description is not None:
            body["description"] = description
        return self._client.post("/integration/custom-instructions", json=body)

    def update(
        self,
        instruction_id: str,
        key: Optional[str] = None,
        value: Optional[str] = None,
        description: Optional[str] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {}
        if key is not None:
            body["key"] = key
        if value is not None:
            body["value"] = value
        if description is not None:
            body["description"] = description
        return self._client.patch(f"/integration/custom-instructions/{instruction_id}", json=body)

    def delete(self, instruction_id: str) -> Any:
        return self._client.delete(f"/integration/custom-instructions/{instruction_id}")

    def suggestions(self) -> List[Dict[str, Any]]:
        return self._client.get("/integration/custom-instructions/suggestions")
